### PR TITLE
DEV-4080 Fixing Subgrant Linking Logic

### PR DESCRIPTION
**Pull request by dpb-bah**
**High level description:**
Updating the FSRS Grant linking logic to be fully accurate with FABS.  

**Technical details:**
* Matching case-insensitive
  * fain (dashes removed) <-> fain
  * federal_agency_id <-> awarding_sub_tier_agency_c 
* Alembic migration for a couple of indexes

**Link to JIRA Ticket:**
[DEV-4080](https://federal-spending-transparency.atlassian.net/browse/DEV-4080)

The following are ALL required for the PR to be merged:
- [ ] Backend review completed
- [x] Unit & integration tests updated with relevant test cases
- [x] Reloaded Subgrants on Sandbox
- Frontend impact assessment completed
- Documentation Updated

### CODE REVIEW

File: dataactcore/models/stagingModels.py [hunk 1]
```diff
@@ -1694,6 +1694,7 @@ def __init__(self, **kwargs):
 Index("ix_pafa_uri_upper", func.upper(PublishedAwardFinancialAssistance.uri))
 Index("ix_pafa_awarding_subtier_c_upper", func.upper(PublishedAwardFinancialAssistance.awarding_sub_tier_agency_c))
 Index("ix_pafa_afa_generated_unique_upper", func.upper(PublishedAwardFinancialAssistance.afa_generated_unique))
+Index("ix_pafa_record_type", func.upper(PublishedAwardFinancialAssistance.record_type))
```
Comment (alburde1) [inline dataactcore/models/stagingModels.py]: 1) Why are we uppering the record type? Why not just set `index=True` in the column definition?
2) It is ALREADY defined that way and this index already exists

File: dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql [hunk 2]
```diff
@@ -34,7 +33,9 @@ aw_pafa AS
         AND EXISTS (
             SELECT 1
             FROM unlinked_subs
-            WHERE UPPER(TRANSLATE(unlinked_subs.award_id, '-', '')) = UPPER(TRANSLATE(pafa.fain, '-', ''))
+            WHERE pafa.record_type != 1
+                AND UPPER(TRANSLATE(unlinked_subs.award_id, '-', '')) = UPPER(TRANSLATE(pafa.fain, '-', ''))
+                AND UPPER(unlinked_subs.awarding_sub_tier_agency_c) IS NOT DISTINCT FROM UPPER(pafa.awarding_sub_tier_agency_c)
```
Comment (alburde1) [inline dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql]: Have you run any tests to see if `IS NOT DISTINCT FROM` is faster or slower than `COALESCE`?

This comment applies to all instances of `IS NOT DISTINCT FROM`

File: dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql [hunk 3]
```diff
@@ -34,7 +33,9 @@ aw_pafa AS
         AND EXISTS (
             SELECT 1
             FROM unlinked_subs
-            WHERE UPPER(TRANSLATE(unlinked_subs.award_id, '-', '')) = UPPER(TRANSLATE(pafa.fain, '-', ''))
+            WHERE pafa.record_type != 1
```
Comment (alburde1) [inline dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql]: Should be using `<>` not `!=`

File: dataactcore/scripts/raw_sql/populate_subaward_table_grants.sql [hunk 4]
```diff
@@ -25,8 +22,10 @@ WITH aw_pafa AS
         AND EXISTS (
             SELECT 1
             FROM fsrs_grant
-            WHERE UPPER(TRANSLATE(fsrs_grant.fain, '-', '')) = UPPER(TRANSLATE(pafa.fain, '-', ''))
+            WHERE record_type != 1
```
Comment (alburde1) [inline dataactcore/scripts/raw_sql/populate_subaward_table_grants.sql]: Should be using `<>` not `!=`

File: dataactcore/models/stagingModels.py (hunk 1 above)
Comment (dpb-bah) [reply dataactcore/models/stagingModels.py]: Force of habit of UPPERCASE'ing everything and confirmed there are no letter record_types in the data. Removing.

File: dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql (hunk 2 above)
Comment (dpb-bah) [reply dataactcore/scripts/raw_sql/link_broken_subaward_grants.sql]: Both are very similar in time with `IS NOT DISTINCT FROM` being just a little faster.
-- 6 rows, 509.5s IS NOT DISTINCT
-- 6 rows, 551.1s COALESCE

**Pull request merged**